app/scripts/migrations/152.ts

```diff
@@ -0,0 +1,313 @@
+  // that have the `endowment:ethereum-provider` permission.
+  const entries = Object.entries(permissionControllerState.subjects) as [
+    string,
+    GenericPermissionControllerSubject,
+  ][];
+
+  permissionControllerState.subjects = entries.reduce<
+    PermissionControllerSubjects<PermissionConstraint>
+  >((accumulator, [key, subject]) => {
+    const permissionKeys = Object.keys(subject.permissions);
+    const needsMigration = permissionKeys.includes(
+      SnapEndowments.EthereumProvider,
+    );
+
+    if (!needsMigration) {
+      return {
+        ...accumulator,
+        [key]: subject,
+      };
+    }
+
+    updatedSubjects.push(key);
+
+    const caip25PermissionCaveatWithCurrentChainIdSet = addPermittedEthChainId(
+      getExistingCaip25PermissionCaveat(subject),
+      currentChainId,
+    );
+
+    const {
+      date = Date.now(),
+      id = nanoid(),
+      invoker = key,
+      parentCapability = Caip25EndowmentPermissionName,
+    } = subject.permissions[Caip25EndowmentPermissionName] ?? {};
+
+    const newSubject: GenericPermissionControllerSubject = {
+      ...subject,
+      permissions: {
+        ...subject.permissions,
+        [Caip25EndowmentPermissionName]: {
+          caveats: [
+            {
+              type: Caip25CaveatType,
+              value: caip25PermissionCaveatWithCurrentChainIdSet,
+            },
+          ],
+          date,
+          id,
+          invoker,
+          parentCapability,
+        },
+      },
+    };
+
+    return {
+      ...accumulator,
+      [key]: newSubject,
+    };
+  }, {});
+
+  const currentDomains = selectedNetworkControllerState.domains;
+  const domains = Object.fromEntries(
+    updatedSubjects.map((subject) => [
+      subject,
+      networkControllerState.selectedNetworkClientId,
+    ]),
+  );
+
+  selectedNetworkControllerState.domains = {
+    ...currentDomains,
+    ...domains,
+  };
+
+  return state;
+}
+
+/**
+ * Get the existing `endowment:caip25` caveat value for a given subject, if any.
+ * If the caveat does not exist, a new empty caveat value is returned.
+ *
+ * This function does not validate the structure of the existing caveat value,
+ * and assumes that the existing value is valid.
+ *
+ * @param subject - The subject to get the existing caveat value for.
+ * @returns The existing `endowment:caip25` caveat value.
+ */
+function getExistingCaip25PermissionCaveat(
+  subject: GenericPermissionControllerSubject,
+): Caip25CaveatValue {
+  const existingCaveat = subject.permissions[
+    Caip25EndowmentPermissionName
+  ]?.caveats?.find((caveat) => caveat.type === Caip25CaveatType);
+
+  if (!existingCaveat) {
+    return {
+      isMultichainOrigin: false,
+      optionalScopes: {},
+      requiredScopes: {},
+    };
+  }
+
+  return existingCaveat.value as Caip25CaveatValue;
+}
+
+function getChainId(
+  networkClientId: string,
+  networkConfigurationsByChainId: RuntimeObject,
```

Comment:
Any reason we can't type this as:
`networkConfigurationsByChainId: Record<string, NetworkConfiguration>` ?